engineering-plans/0002-ethereum-tracing-cache.md

```diff
@@ -0,0 +1,184 @@
+
+We want to minimize the cost of external requests for trace data. It is likely that it is better too...
+ * Make fewer requests
+ * Not ask for trace data that is already cached
+ * Ask for trace data for multiple contract addresses within the same block when possible.
+
+ There is one flow of data which upholds these invariants. In doing so it makes a tradeoff of increasing latency for the execution of a specific subgraph, but increases throughput of the whole system.
+
+ Within this graph:
+ * Edges which are labelled refer to some subset of the output data.
+ * Edges which are not labelled refer to the entire set of the output data.
+ * Each node executes once for each contiguous range it. That is, it merges all incoming data before executing, and executes the minimum possible times.
+ * The example given is just for 2 addresses. The actual code must work on sets of addresses.
+
+ ```mermaid
+ graph LR;
+    A[Block Range for Contract A & B]
+    A --> |Above Reorg Threshold| E
+    D[Get Cache A]
+    A --> |Below Reorg Threshold A| D
+    A --> |Below Reorg Threshold B| H
+    E[Ethereum A & B]
+    F[Ethereum A]
+    G[Ethereum B]
+    H[Get Cache B]
+    D --> |Found| M
+    H --> |Found| M
+    M[Result]
+    D --> |Missing| N
+    H --> |Missing| N
+    N[Overlap]
+    N --> |A & B| E
+    N --> |A| F
+    N --> |B| G
+    E --> M
+    K[Set Cache A]
+    L[Set Cache B]
+    E --> |B Below Reorg Threshold| L
+    E --> |A Below Reorg Threshold| K
+    F --> K
+    G --> L
+    F --> M
+    G --> M
+ ```
+
+ 
+This construction is designed to make the fewest number of the most efficient calls possible. It is not as complicated as it looks. The actual construction can be expressed as sequential steps with a set of filters preceding eah step.
+
+### Useful dependencies
+The feature deals a lot with ranges and sets. Operations like sum, subtract, merge, and find overlapping are used frequently. [nested_intervals](https://crates.io/crates/nested_intervals) is a crate which provides some of these operations.
+
+
+## Tests
+
+### Benchmark
+A temporary benchmark will be added for indexing a simple subgraph which uses call handlers. The benchmark will be run in these scenarios:
+* Sync before changes
+* Re-sync before changes 
+* Sync after changes
+* Re-sync after changes
+
+### Ranges
+Due to the complexity of the resource minimizing data workflow, it will be useful to have mocks for the cache and database which record their calls, and check that expected calls are made for tricky data sets.
+
+### Database
+A real database integration test will be added to test the add/remove from cache implementation to verify that it correctly merges blocks, handles concurrency issues, etc.
```

Comment:
Isn't clear to me yet what the concurrency risks are. Is the main problem to deal with that things may change between the time the valid cache range for an address (or no address) is determined and the actual traces are fetched from the cache?

Thinking about this makes me a little concerned about the amount of synchronization that'll be necessary between subgraph-indexing threads to ensure there are no race conditions. @lutter what do you think?

Reply:
The concern with concurrency is that two tables are being utilized which need to preserve atomicity (the table for the actual blocks, and the table for the cache). Because multiple subgraphs may include the same data sources and be indexed at the same time, they contend for the db.

I'm confident the use of serialized transactions as suggested by lutter will do the trick. But, it's worth making sure that the additional code paths introduced by serialized transactions (mainly retry failed transactions) are in fact tested because it may be some time before the code paths would be hit otherwise.